maestrowf/utils.py

```diff
@@ -280,6 +284,273 @@ def create_dictionary(list_keyvalues, token=":"):
     return _dict
 
 
+def next_path(path_pattern):
+    """
+    Finds the next free path in an sequentially named list of files
+
+    e.g. path_pattern = 'file-%s.txt':
+
+    file-1.txt
+    file-2.txt
+    file-3.txt
+
+    Runs in log(n) time where n is the number of existing files in sequence
+    https://stackoverflow.com/questions/17984809/how-do-i-create-a-incrementing-filename-in-python
+    """
+    i = 1
+
+    # First do an exponential search
+    while os.path.exists(path_pattern % i):
+        i = i * 2
+
+    # Result lies somewhere in the interval (i/2..i]
+    # We call this interval (a..b] and narrow it down until a + 1 = b
+    a, b = (i // 2, i)
+    while a + 1 < b:
+        c = (a + b) // 2  # interval midpoint
+        a, b = (c, b) if os.path.exists(path_pattern % c) else (a, c)
+
+    return path_pattern % b
+
+
+def splitall(path):
+    """
+    https://www.oreilly.com/library/view/python-cookbook/0596001673/ch04s16.html
+    """
+    allparts = []
+    while 1:
+        parts = os.path.split(path)
+        if parts[0] == path:  # sentinel for absolute paths
+            allparts.insert(0, parts[0])
+            break
+        elif parts[1] == path:  # sentinel for relative paths
+            allparts.insert(0, parts[1])
+            break
+        else:
+            path = parts[0]
+            allparts.insert(0, parts[1])
+    return allparts
+
+
+def recursive_render(tpl, values):
+    """
+    https://stackoverflow.com/questions/8862731/jinja-nested-rendering-on-variable-content
+    """
+    prev = tpl
+    while True:
+        curr = Template(prev).render(values)
+        if curr != prev:
+            prev = curr
+        else:
+            return curr
+
+
+def valid_link_template(astring):
+    if astring is None:
+        return None
+    error_text = (
+            "\nTemplate error: '" + astring + "'\n"
+            "    does not include required substrings "
+            "{{instance}} and {{step}}.\n")
+    if (astring.find('{{instance}}') == -1
+       or astring.find('{{step}}') == -1):
+        print(error_text)
+        raise ValueError(error_text)
+    return astring
+
+
+class Linker:
```

Comment:
I tried to rebase this on current develop but ran into an issue where the `{{instance}}` would still get its parameterized name instead of a hash. In the current develop, I introduced some new properties to the `_StepRecord` class that conflicts with your changes, plus I believe it also has a `real_name` property that provides what I think `step_label` is trying to provide.

Do you mind rebasing on the current develop? I tried to debug it and make the necessary changes, but I had a hard time figuring out exactly what was wrong. In that vain, could you comment this class and some of its larger methods? I found it difficult to infer what the intent of things were in a number of places.